pkg/tracing/tracing.go

```diff
@@ -0,0 +1,157 @@
+package tracing
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/siglens/siglens/pkg/config"
+	log "github.com/sirupsen/logrus"
+	"github.com/valyala/fasthttp"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/sdk/resource"
+	"go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
+	traceNoop "go.opentelemetry.io/otel/trace/noop"
+)
+
+type customExporter struct {
+	exporter     *otlptrace.Exporter
+	failureCount int
+	lock         sync.Mutex
+}
+
+// Shutdown implements trace.SpanExporter.
+func (ce *customExporter) Shutdown(ctx context.Context) error {
+	log.Info("Shutting down the exporter")
+	return ce.exporter.Shutdown(ctx)
+}
+
+func (ce *customExporter) ExportSpans(ctx context.Context, spans []trace.ReadOnlySpan) error {
+	err := ce.exporter.ExportSpans(ctx, spans)
+	ce.lock.Lock()
+	defer ce.lock.Unlock()
+
+	if err != nil {
+		ce.failureCount++
+		log.Errorf("Traces export failed: %v", err)
+
+		// Check if the failure threshold is exceeded.
+		if ce.failureCount >= 3 {
+			fmt.Println("Disabling tracing due to repeated export failures.")
+			log.Errorf("Disabling tracing due to repeated export failures.")
+			config.SetTracingEnabled(false)
+
+			// Set the tracer provider to a no-op provider. This will prevent any further spans from being exported.
+			otel.SetTracerProvider(traceNoop.NewTracerProvider())
+
+			// Shut down the custom exporter to clean up resources.
+			if shutdownErr := ce.Shutdown(ctx); shutdownErr != nil {
+				log.Errorf("Failed to shut down the tracer provider: %v", shutdownErr)
+			}
+		}
+	} else {
+		// Reset failure count on a successful export.
+		ce.failureCount = 0
+	}
+
+	return err
+}
+
+// InitTracing initializes the OpenTelemetry tracing.
+func InitTracing(serviceName string) func() {
+
+	if !config.IsTracingEnabled() {
+		log.Info("Tracing is disabled")
+		return func() {}
+	}
+
+	if serviceName == "" {
+		log.Errorf("Service name is required to initialize tracing")
```

Comment:
nit: better error msg `Service name was not provided in config, assuming a default service name as siglens`